### ISSUE TITLE
Registering functions and types to the engine at runtime

### ISSUE THREAD

**Issue by limads** [rhaiscript/rhai#494]
# Motivation

Suppose I want my users to be able to load native Rust functions into the engine at runtime (I'm writing a REPL), by using something like libloading to resolve a symbol from a C dynamic library (.so) compiled by the user:

User code:

```rust
// I want the user to be able to load this at runtime.
fn user_func(a : i64) -> i64 { a };

// The user would write something like:
#[no_mangle]
extern "C" fn register_user_func(engine : &mut rhai::Engine) {
	engine.register_fn("user_func", |a : i64| -> i64 { user_func(a) });
}
```

This assumes the user crate depends on the same rhai version and is compiled with the same compiler version as the host REPL is compiled with, to guarantee the types will have the same memory layout. 

The code to load the function at the REPL is something like:

```rust
type RegisterSymbol = libloading::Symbol<unsafe extern fn(*mut rhai::Engine)>;

let mut engine = Engine::new();
let lib = libloading::Library::new("path/to/user/crate/target/release/libuser.so").unwrap();
let register_func : RegisterSymbol = lib.get(b"register_user_func").unwrap();
unsafe { register_func(&mut engine as *mut _) };
println!("{}", engine.eval::<i64>("user_func(1)"));
```

The advantage here is that the user does not need to worry about details of how the engine is being managed. The "host" REPL is compiled once (an might be distributed, for example, with a cargo install) and the user just needs to worry about exposing his implementation of a standard registration API.

Currrently, this won't work because the ahash::AHasher will generate different function hashing strategies at the version compiled by the user, and the version compiled by the host. Any functions registered from a dynamic library loaded at runtime will not be found by the Engine (from a quick investigation, it seems there is an RNG seed generated at compile time that is used to determine the function hash, so this is expected).

# Proposed solution

By using constant keys when generating the hasher at `src/func/hashing.rs`, the hashing at the client and the host REPL will be the same, and the strategy I'm trying above works.

```rust
/// Create an instance of the default hasher.
#[inline(always)]
#[must_use]
pub fn get_hasher() -> ahash::AHasher {
    //Default::default() < -- Replace this
    ahash::AHasher::new_with_keys(1234, 5678) // < -- With this (or any other fixed key)
}
```

I'm proposing to allow this user case by, for example, adding a custom feature that uses a fixed hash, making the function hashes compatible across compilations:

```rust
/// Create an instance of the default hasher.
#[inline(always)]
#[must_use]
pub fn get_hasher() -> ahash::AHasher {

    #[cfg(feature="constant_hasher")
    return ahash::AHasher::new_with_keys(1234, 5678);

    Default::default()
}
```

There might be security issues involved with using the strategy above I'm not aware of, which is why I think this use
case could be solved via a custom feature not enabled by default.

**Comment by schungx** [comment]
There doesn't seem to be such a limitation right now.  Rhai can already load functions during runtime.  Hashes are recalculated during the `Engine::register_fn` call, so the external function does not have to worry about hashing.

To support externally-loaded functions with variable parameter types, you may use `Engine::register_raw_fn` which lets you specify the types of parameters.  However, that is where the real complication arises.

Rhai relies on `std::any::TypeId` to generate a unique hash ID for each type during compile time.  When you dynamically load an external library, it is quite impossible to generate `TypeId`'s for parameter types, unless they are pre-generated inside that library.

It is understood that `TypeId` is not stable between compiler versions and possibly different environments. This is the main issue with getting what you are trying to do to work -- mismatch between the external library's `TypeId`'s and Rhai's.

**Comment by schungx** [comment]
What you need to register a function with Rhai are several things:

1) The name of the function, which should be easy
2) A list of parameter types, in `TypeId` format -- that's the potential incompatibility, especially when the external function is in a library not written in Rust (e.g. written in C)
3) A pointer to that function, as a trait object like `dyn FnAny` -- since you're loading the library as an external C library, the function is in C calling convention, not Rust.  Therefore, you may not be able to make it a function trait object to pass to Rhai.

In other words, the main killer would be the incompatible calling convention between a Rust function and a C function, and you won't be able to avoid that if you load the external library as a C lib.

**Comment by schungx** [comment]
OK, the first Google result came back with this: https://rust-lang.github.io/unsafe-code-guidelines/layout/function-pointers.html

It seems that the C and Rust function ABI's are compatible.  In this case you can make this work.

**Comment by limads** [comment]
Thanks for replying in such short notice.

Yes, the `TypeId` is one issue with this solution. But I'm assuming for now that the external library I want to call will always be compiled from Rust (supporting C libraries is not a goal), using the same compiler version used to compile the executable of the REPL I'm writing. This can be enforced by requiring the user API to implement a call to `rustc_version::version()`. This can be done, e.g. by requiring the user to implement a trait, which calls this function from a provided method, so the user does not have to worry about it: Just throw a runtime error for any version mismatches before loading the lib. The actual function pointers passed via `engine.register_fn` are always valid while the REPL is running, since the libraries I'm loading always outlive the rhai Engine. 

My objective is not register into the engine the loaded symbols per se - But to just recover a single function symbol from each user library, that does all the registration. The user will have `rhai` as a dependency, and call `register_fn` from his own function implementation. When that happens, the function will receive a hash. The `Default` implementation for `AHasher` that `rhai` uses relies in the following static variable located at `aHash/src/random_state.rs` (Line 49):

```rust
#[cfg(not(all(target_arch = "arm", target_os = "none")))]
static RAND_SOURCE: OnceBox<Box<dyn RandomSource + Send + Sync>> = OnceBox::new();
```

Here is where the issue happens: The user compiled his lib with one static copy of the variable, but my REPL (that instantiated the engine) uses another copy of the same static variable, so the hashes will be different, and rhai will be unable to find the function, returning an `EvalAltResult::ErrorFunctionNotFound`. If instead of using the `Default` implementation for `AHasher`, we call `AHasher::new_with_keys` (aHash/src/aes_hash.rs` (Line 53) as in the snippet I provided above, there is no dependency on any static variables when creating the `AHasher`, and `rhai` is able to load the user function successfully.

**Comment by schungx** [comment]
Ah.  Now I get it.  You'd need a hasher with predictable output, such as from a static seed.

I believe `AHasher` supports such a thing... it may be easy to add to Rhai, as all hashing boils down to one single function to get a hasher.

**Comment by limads** [comment]
Exactly. If you agree to implement this as a crate feature I can submit a PR with the implementation I described above. There are other ways to do that as well (for example, with an `Engine::new_with_hasher` method), but implementing as an optional crate feature wouldn't require any changes in the API, or require the engine to have one extra field to store a custom hasher.

**Comment by schungx** [comment]
Yes, I suppose this usage is rare enough that making it into the API doesn't make sense.  Better have a feature gate.

Would appreciate a PR!  Hopefully you can demonstrate that it works in real life...

**Comment by schungx** [comment]
Come to think of it...  I am not sure why `AHasher` would need to be called both in the library and in the `Engine`... the only place where hashing is performed is all done in the `Engine`...

The only place where such hashes are calculated outside of the `Engine` is when you create a module (which requires hashing the functions).

So if you define modules in the library, then yes, both hashers need to be the same.  If you use `register_raw_fn` exclusively, then you won't need to have two hashers.

**Comment by limads** [comment]
I'm not worrying about modules yet. My objective, for now, is just to load types and functions using `register_type` and `register_fn` to the global namespace, but called from possibly many differently dynamic libraries compiled from different user crates (I'd like to avoid the raw function API, since `register_fn` is much simpler to use, since the user can just forward plain Rust functions to it without worrying about determining TypeIds manually). I fail to see how the issue can solved with the raw function API, however. Suppose the user calls `register_raw_fn` instead of `register_fn`. This retrieves the global namespace module and eventually calculates a hash at `set_fn` (`src/module/mod.rs:705`):

```rust
let hash_fn = calc_native_fn_hash(empty(), name.as_ref(), &param_types);
```

Which is stored at the `functions` field of the module.

Now when the program evaluates any expression, it will hit another call requiring the hasher at the parsing stage, such as in `parse_stmt` (src/parser.rs:2815):

```rust
let hash = calc_fn_hash(&func.name, func.params.len());
```

So the problem still persists even if the user uses uses the `register_raw_fn` API from its library, that is, if I'm not missing some detail about the implementation. `FnCallExpr` always uses functions hashes, right?

**Comment by schungx** [comment]
`register_fn` internally calls `register_raw_fn` so I believe it is simply syntactical sugar.

Yes, function hashes are pre-calculated during parsing, but usually you parse the script with the same `Engine`, right?  No?

If you parse the script with the same `Engine` which you did the `register_fn`, then it'll use the exact same hasher.

Of course, you can compile the script with a different `Engine` from a different library (why would you want to?), in that case yes, you'd need stability of the hash.

I suggest you simply do it and see if it works as it stands.  You can debug-print the `AST` to check out the hashes.  If you want, I can include the function hashes in the JSON metadata. Then it would be easy for you to debug.

**Comment by limads** [comment]
Yes, the engine to which functions are registered is the same engine that does the parsing. I think that storing an empty `ahash::AHasher` as a field of the Engine (cloned each time it is required) is another way to solve the issue, since it would behave according to your description. But in the current implementation, it seems `parse_fn_call` makes calls to `calc_fn_hash` and `calc_qualified_fn_hash`, which in turn creates a new instance of `ahash::AHasher` via its `Default` implementation every time a `rhai` expression containing function calls is parsed. You can assume this hasher behaves in the same way when all functions are registered and used from the same compiled crate, but you run into the duplicate definition of the static random key I mentioned above if they are registered from crates compiled separately. 

I am currently able to register functions from a dynamically loaded library by depending on a local version of rhai where I made the change to a fixed hasher I described initially (although only for primitive types such as i64 and f64, I'm still running into some TypeId mismatches for String functions).

**Comment by schungx** [comment]
In that case, maybe the solution is, as you said, to create a field that holds a `AHasher` with an (possibly `internals`) API `Engine::new_with_hasher` and/or `Engine::new_raw_with_hasher`.

**Comment by schungx** [comment]
After some experimentation, it would not work to add a `hasher` field to `Engine` unless I add one to the entire API for `Module` as well, since it is entirely possible that modules are built outside of any `Engine`.

How about going the other way and providing `Engine::get_hasher` which returns an instance of the `AHasher` used by the `Engine`?

**Comment by limads** [comment]
The way I see it the feature gate is the least disruptive way of allowing functions to be registered across compilation units. There might problems of doing that in a security-sensitive context (which I don't have any experience with), so perhaps the feature, if implemented, could be named `untrusted_hasher` or something like that.

Today I came up with a different approach to solve the problem that does not require the client to register the functions, therefore not requiring any changes to the hashing logic as it is implemented. Instead of doing the registration on the user-compiled dynamic library, I do it on the host, by requiring the user to just hand me the information required to register all his functions (packed into a boxed type for FFI compatibility):

```rust
// User will implement something like this...
pub struct FuncInfo {
    name : String,
    args : Vec<String>,
    func : CallableFunction
}

#[no_mangle]
pub extern "C" fn get_funcs() -> Box<Vec<FuncInfo>> {
    // User passes his functions here
}
```

```rust
// And the host registers them with
let mut module = rhai::Module::new();
for func in *get_funcs() {
  // Keep TypeIds of host in a parallel registry  (a HashMap<String, TypeId>, 
  // mapping namespace-qualified types to ids)
  let param_types : Vec<TypeId> = func.args.iter().map(|ty| { host_types[&ty[..]] }).collect();
  module.set_fn(
      func.name,
      rhai::FnNamespace::Global,
      rhai::FnAccess::Public,
      Some(&func.types.iter().map(|v| v.as_ref()).collect::<Vec<_>>()),
      &param_types[..],
      func.func,
  );
}
engine.register_global_module(module.into());
```

I can confirm this works to register functions that work with basic `rhai` types. I've managed to write a generic wrapper similar to what `set_native_fn` does, but to generate this intermediary structure instead of registering the function directly. I haven't tested custom types yet, but since registering them seems to just add their name to a catalog kept by the engine, it can be done on the host side as well. I'm still studying how to register getters, setters and iterators using this method.

It seems the library had the tools that I required to solve this problem (`CallableFunction` and `RegisterNativeFunction::into_callable_function`), but this solution wasn't clear to me at first, registering the types at the client library seem to be the most natural solution. If support for dynamically-loaded modules is not planned for `rhai`, perhaps this topic is a good candidate for at least a new documentation section.

**Comment by schungx** [comment]
> Instead of doing the registration on the user-compiled dynamic library, I do it on the host, by requiring the user to just hand me the information required to register all his functions (packed into a boxed type for FFI compatibility):

I think this is the best way to approach it.

> If support for dynamically-loaded modules is not planned for `rhai`, perhaps this topic is a good candidate for at least a new documentation section.

In my personal project (that uses Rhai), I actually load dynamic libraries, but I didn't do it like you did by wrapping a `CallableFunction` across FFI.  I recreated the Rust API on the host side.  Your solution seems much easier and versatile except for the fact that we must use the same version of Rust to compile both (`TypeId` stability).  This may not work out for my requirements, but some applications may be able to live within this restriction.

I strongly suggest you have a working example up, and I can write up a new "patterns" topic for the Book.

**Comment by limads** [comment]
Relying on `TypeId` might be an issue even using the same compiler version. One issue I ran across had to to with `ImmutableString`. I could get the `TypeId` of `i64`, `f64` `bool` (all from standard library) to match between host and client. But the `TypeId` of `ImmutableString` did not match (even with the same compiler version, and depending on the same local `rhai` version). I suspect that it is because `ImmutableString` relies on a third-party crate for its implementation (`SmartString`), and the compiler might be doing something different to resolve the `TypeId` across builds. Since `rhai` converts the `TypeId` of any `&str` and `String` arguments to `ImmutableString` internally, any functions relying on those also didn't work. That was the motivation of passing the type names into my `FuncInfo` structure: I just match the qualified type names recovered at the client with `RegisterNativeFunction::param_names()` to the `TypeId` at the host. Since `rhai` relies on just a few types and I'm not anticipating any collision issues.

I do plan on getting my solution published (it will most likely be a REPL offered with a cargo install), but I have to figure out some things first. For the host to have feature-parity with the generic functions that `rhai` offer (`register_type_with_name`, `set_getter_fn`, `set_iterator`, etc), I would need non-generic versions of those (`register_raw_type_with_name`, `set_raw_getter_fn`, `set_raw_iterator`, etc) that took the `std::any::type_name()` and `std::any::TypeId::of` of the user types. It is not hard to duplicate the implementations for how, since they all call `Modue::set_fn internally`, but having them implemented at `rhai` itself (perhaps behind the `internals` feature) would minimize the risk of incompatibilities at future updates. I'll share the implementation once I have it working.

**Comment by schungx** [comment]
If you're only facing a mismatch for `ImmutableString`, maybe we can track down the core problem?  For example, display the `TypeId` both in Rhai as well as in the external library.

Try with other structs to see if it is a struct issue, or at the core a `SmartString` issue.

To be honest, I can't think of a reason why `TypeId` would be portable for some types and not others... so there must be a deeper reason at work here.

**Comment by schungx** [comment]
I checked on my machine and Rust seems to generate the same `TypeId` for `ImmutableString` every time...

One of the potential reasons is perhaps two independent copies of `SmartString` being compiled in by `cargo` with the Rhai host and the library both using different versions of the same `SmartString`?

I'd definitely suggest trying the `TypeId` for, say, `Vec` and some other `struct`s to see if you get a consistent behavior here.

**Comment by schungx** [comment]
OK.  Gotcha.  I think I know why.

`TypeId` seems to include the _entire path_ of the type in its hash.

That's why primitive types like `u64` are the same.  They are called the same in all places.

For `ImmutableString`, the Rhai `Engine` and  your library see differently:

`Engine`:  I see `crate::ImmutableString`

Your library: I see `rhai::ImmutableString`

So the names of non-primitive types are different because their roots are different.

**Comment by limads** [comment]
I think you are right about the namespace difference. Relying on the `std::any::type_name()` seems to be a safe bet for now, since fully qualified namespaces won't change between client and user. I'm now using the client type names as indices to a `HashMap` (built at the host) which hold the host `TypeId` for primitive and rhai types. Any time I register a new user type, I have to register the client `TypeId` as well, since the host has no way of knowing what types exist or not. Since the hashes seems to be based on the type namespace, I don't think duplicates will be a problem.

I've now managed to replicate everything that `rhai` does with generic registering functions, but to generate the intermediate `FuncInfo` that is handed to the host. Methods, indexers, iterators, all looking good. The only change I had to make to `rhai` was to add a non-generic way to register types:

```rust
impl Engine {

// ...

/// Registers a type to the engine that is not actually represented in the current 
/// program (e.g. a type that exists only behind a dynamic library loaded at runtime).
/// Qualified is the output of std::any::type_name(). short is the actual type name 
/// (used to build the function metadata)
pub fn register_raw_type_with_name(&mut self, qualified : &str, short: &str) -> &mut Self {
    self.type_names.insert(qualified.into(), Box::new(short.into()));
    self
}

// ...

}
```
Any chance of adding this to the library? This is the only missing piece I need.

**Comment by schungx** [comment]
> One extremely simple solution is to simply replace `TypeId::of::<ImmutableString>()` with a pre-generated constant, thus both sides can use it.

Come to think about it, even simpler would be to have a new `internals` only API that simply returns the `TypeId` of `ImmutableString` as a value so outside crates can use it.

> Any chance of adding this to the library? This is the only missing piece I need.

That would be no problem at all.  I'll add that under the `internals` gate.

**Comment by limads** [comment]
Great! Give me a few days to check that everything is working, and if so I'll close this issue.

**Comment by limads** [comment]
I can confirm my solution is working with the 1.3 version of the crate, since it is now possible to register types to the engine without knowing what the types are at compile time (via `Engine::register_type_with_name_raw`). For future reference, the best way to solve the issue was to send the `rhai::plugin::CallableFunction` over the FFI boundary (easily obtained via `RegisterNativeFunction::into_callable_function`). At the host engine, the user function can then be registered via `Module::set_fn`. Doing that avoids both the hashing issue, and the possible difference between the `TypeId` of ImmutableString at the host and client REPL. Thanks for the help, hopefully I'll have something up on crates.io soon.
